### PR TITLE
Switch to mypy part1

### PR DESCRIPTION
**Pull request by m-vdb**
**Proposed changes**:
Fix the following typing issues:
- `dict-item`
- `list-item`
- `call-arg`

**Leftovers to merge this PR**
- [x] [this discussion in `rasa.core.actions.action`](https://github.com/RasaHQ/rasa/pull/6511#discussion_r503945014) to resolve _(needs input from reviewer)_
- [x] [this discussion in `rasa.core.actions.two_stage_fallback`](https://github.com/RasaHQ/rasa/pull/6511#discussion_r503948950) to resolve _(needs input from reviewer)_
- [x] [this discussion in `rasa.core.test`](https://github.com/RasaHQ/rasa/pull/6511#discussion_r503955238) to resolve _(ready to implement)_
- [x] [this discussion in `rasa.shared.nlu.training_data.lookup_tables_parser`](https://github.com/RasaHQ/rasa/pull/6511#discussion_r503967424) to resolve _(ready to implement)_
- [x] [this discussion on `tests`](https://github.com/RasaHQ/rasa/pull/6511#discussion_r503969854) to resolve _(needs input from reviewer)_
- [x] API spec to update (see `rasa.server` module)
- [x] merge `main` in the branch and fix merge conflicts
- [ ] ~Update `rasa.model.Fingerprint` typing (possibly in another task)~

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by m-vdb** [comment]
@wochinge after merging master into this branch, I have to fix more issues. I have a question about [`ActionExecuted`](https://github.com/RasaHQ/rasa/blob/056f6bd83fba1458559e67a941aaea00189b9dc9/rasa/shared/core/events.py#L1031-L1055). Is it safe to assume that it should only be initialised with `action_name` and/or `action_text`, but if both are `None`, then we should raise an error?

File: rasa/nlu/model.py
Comment (wochinge) [inline rasa/nlu/model.py:87]:
```suggestion
    def __init__(self, metadata: Dict[Text, Any]) -> None:
```

File: rasa/nlu/model.py
Comment (wochinge) [inline rasa/nlu/model.py:396]: seems we should rather pass an `int` here instead of `Text`. The only usages of `time` I found, cast to an `int` 🙈

File: rasa/nlu/model.py
Comment (wochinge) [inline rasa/nlu/model.py:205]: omg 🤦

File: tests/cli/test_rasa_interactive.py
Comment (wochinge) [inline tests/cli/test_rasa_interactive.py:65]: This change is wrong. It mocks the actual `train` function from the `rasa.train` module. 
It was not the best idea back then to have functions with the same name as modules on the same level 😬

File: rasa/server.py
Comment (wochinge) [inline rasa/server.py:248]: Rick is currently removing this in https://github.com/RasaHQ/rasa/pull/6979

**Review by wochinge — COMMENTED** [review]
All done besides from a small snippet in `rasa.server`.

Nice work! Cool to see that mypy also helps us detecting and removing technical debt! 🥇

File: rasa/server.py
Comment (wochinge) [inline rasa/server.py:142]: why do we need to define a `SanicView` here? (that concludes my review)

**Comment by m-vdb** [comment]
@wochinge thanks for your review! This PR was not really ready for review, but [the "part0" is](https://github.com/RasaHQ/rasa/pull/6470). Would you mind taking a look?

**Comment by wochinge** [comment]
🙈  will do

File: rasa/cli/train.py
Comment (m-vdb) [reply rasa/cli/train.py:10]: dammit, it's really misleading.

File: rasa/core/channels/rasa_chat.py
Comment (m-vdb) [reply rasa/core/channels/rasa_chat.py:76]: usually python standard library functions raise a TypeError for this kind of errors (`None` is of type `NoneType`)

File: rasa/nlu/model.py
Comment (m-vdb) [reply rasa/nlu/model.py:85]: It's not really needed. It's only needed if your init method doesn't have any arguments and you still want it to be type checked, as far as I recall. And also, `-> NoReturn` would be more accurate anyway

File: rasa/nlu/model.py
Comment (m-vdb) [reply rasa/nlu/model.py:394]: I don't have enough knowledge about this to make the decision myself, but don't you think it would be a breaking change that could impact users?

File: rasa/server.py
Comment (m-vdb) [reply rasa/server.py:248]: well, in that case, we'll simply handle the merge in either of the PRs, depending on which one gets merged first :)

File: tests/cli/test_rasa_interactive.py
Comment (m-vdb) [reply tests/cli/test_rasa_interactive.py:65]: I'm confused, what should I do here?

File: rasa/core/actions/two_stage_fallback.py
Comment (m-vdb) [reply rasa/core/actions/two_stage_fallback.py:197]: I've added the type hint on the tracker, but there still an issue to fix to remove the casting: in case `latest_message` is empty, what should be done? Raise? Return `[ActionExecuted(ACTION_LISTEN_NAME)]`?

File: rasa/core/channels/rasa_chat.py
Comment (m-vdb) [reply rasa/core/channels/rasa_chat.py:97]: I think that in this case we should even report it to Sentry, because it's an unexpected exception.

File: rasa/core/test.py
Comment (m-vdb) [reply rasa/core/test.py:360]: I think it will make it fail graciously, see `has_prediction_target_mismatch()`. Could you detail a bit more?

File: pyproject.toml
Comment (m-vdb) [reply pyproject.toml:156]: Not at the moment. I wouldn't add all the type stubs unless we need them. If, when enabling more type checks, we realise that we need it, then we should add it

File: rasa/core/actions/action.py
Comment (m-vdb) [reply rasa/core/actions/action.py:726]: I'd go with an exception as well; what type of error should it be? do you think it should land in Sentry (I think it should)

File: rasa/model.py
Comment (m-vdb) [inline rasa/model.py:39]: I think it should be a TypedDict or a dataclass, not sure how hard it would be to make the change

File: rasa/shared/core/training_data/structures.py
Comment (m-vdb) [inline rasa/shared/core/training_data/structures.py:153]: I think we should address this

File: rasa/shared/nlu/training_data/lookup_tables_parser.py
Comment (m-vdb) [reply rasa/shared/nlu/training_data/lookup_tables_parser.py:22]: 🤔 it sounds like a wrong type annotation then. Will see...

File: rasa/server.py
Comment (m-vdb) [reply rasa/server.py:153]: since we're manipulating arguments inside the decorator and call `f` with `request`, we need the type annotation to tell mypy that `f` is a function that accepts `request` as first argument. Does that make sense?

**Comment by m-vdb** [comment]
@wochinge I finally took the time to come back to this PR. I think that I answered all your comments (with sometimes more questions). When you have the time, I'd appreciate a second round of feedback (no rush at all please)

**Comment by wochinge** [comment]
Will have a look early next week then 👍

File: rasa/core/channels/rasa_chat.py
Comment (wochinge) [inline rasa/core/channels/rasa_chat.py:76]: ?
```suggestion
            raise TypeError("JWT public key is `None`. This is likely caused by an error when retrieving the public key from Rasa X.")
```

File: rasa/core/channels/rasa_chat.py
Comment (wochinge) [reply rasa/core/channels/rasa_chat.py:97]: Sounds good. Then let's include `raise e`?

File: rasa/core/featurizers/single_state_featurizer.py
Comment (wochinge) [inline rasa/core/featurizers/single_state_featurizer.py:63]: what is the not type safe part about it?

File: rasa/model.py
Comment (wochinge) [reply rasa/model.py:39]: Should be fairly simple 👍

File: rasa/nlu/model.py
Comment (wochinge) [reply rasa/nlu/model.py:85]: https://www.python.org/dev/peps/pep-0484/ says

> Note that the return type of __init__ ought to be annotated with -> None

I think it would be more consistent to annotate it everywhere with `None`.

File: rasa/nlu/model.py
Comment (wochinge) [reply rasa/nlu/model.py:394]: I think it's safe to change. `UserUttered` uses a separate `float` timestamp anyway which means users shouldn't see this timestamp ever.

File: rasa/shared/core/events.py
Comment (wochinge) [inline rasa/shared/core/events.py:1215]: Why is this a problem? We return a `Dict[Text, Optional[Text]]` either way

File: rasa/shared/core/trackers.py
Comment (wochinge) [inline rasa/shared/core/trackers.py:21]:
```suggestion
```

File: rasa/shared/core/trackers.py
Comment (wochinge) [inline rasa/shared/core/trackers.py:202]: even more precise
```suggestion
        self.latest_message: Optional[UserUttered] = None
```

File: rasa/shared/core/training_data/structures.py
Comment (wochinge) [inline rasa/shared/core/training_data/structures.py:153]: Can we change `_or_string(story_step_element: List[Event]` to `_or_string(story_step_element: List[UserUttered]`?

File: rasa/shared/nlu/training_data/lookup_tables_parser.py
Comment (wochinge) [reply rasa/shared/nlu/training_data/lookup_tables_parser.py:22]: did you already revisit this one?

File: rasa/server.py
Comment (wochinge) [inline rasa/server.py:1253]: 🚀

File: rasa/server.py
Comment (wochinge) [inline rasa/server.py:981]: can you please update the api spec acoordingly?

File: rasa/server.py
Comment (wochinge) [reply rasa/server.py:153]: Yes, it does. And something like `Callable[[Request, ??], BaseHTTPResponse]` doesn't work as we only know the first argument, but not the others, right?

File: rasa/core/test.py
Comment (wochinge) [reply rasa/core/test.py:366]: If we add only one of them (either `intent_gold` or `predicted_intent`) then the lists `intent_predictions` / `intent_targets` would end up having different lengths. `has_prediction_target_mismatch` would for sure work. However [this](https://github.com/RasaHQ/rasa/blob/1b9e44c1215769b9e7552f7382d347ed04f7f13c/rasa/core/test.py#L276) access by index and would then fail. Probably enough to make the linked line fail safe then.

**Review by wochinge — COMMENTED** [review]
A few more questions, but looks great! I comment for now but should be a fast final review.
Thanks for working on this 🙌

File: rasa/nlu/model.py
Comment (m-vdb) [reply rasa/nlu/model.py:85]: Directly quoting the boss -> https://github.com/python/mypy/issues/3358#issuecomment-465783430 (doesn't happen every day ahah)

File: rasa/shared/nlu/training_data/lookup_tables_parser.py
Comment (m-vdb) [reply rasa/shared/nlu/training_data/lookup_tables_parser.py:22]: no I haven't implemented it yet, wanted to wait for an answer on other points first 👍🏻 will do

File: rasa/server.py
Comment (m-vdb) [reply rasa/server.py:153]: 🤔 will try `Callable[[Request, ...], BaseHTTPResponse]`

Comment (m-vdb) [reply rasa/server.py:151]: I was able to make it work with 297c1dcca4a (it's admittedly simpler)

File: rasa/core/actions/action.py
Comment (m-vdb) [reply rasa/core/actions/action.py:726]: ping @wochinge do you have an opinion on this one?

File: rasa/core/channels/rasa_chat.py
Comment (m-vdb) [reply rasa/core/channels/rasa_chat.py:97]: if we raise, then `logging.exception` becomes useless; will remove

File: rasa/shared/core/events.py
Comment (m-vdb) [reply rasa/shared/core/events.py:1215]: I think it's a problem because an `ActionExecuted` instance should either have `action_text` or `action_name` defined (not both at the same time nor none), as I understand it (let me know if this is wrong).

I added the cast because the type definition of this function and others (including `DialogueStateTracker.set_latest_action()`) expect the type to be `Dict[Text, Text]`, not `Dict[Text, Optional[Text]]`.

File: rasa/shared/core/training_data/structures.py
Comment (m-vdb) [reply rasa/shared/core/training_data/structures.py]: I guess I didn't see that we explicitly check above that every element in the list if of type `UserUttered`. It's a `mypy` issue, because it doesn't infer the type correctly after the type guard. I've adapted using ab748f6307a

File: rasa/core/featurizers/single_state_featurizer.py
Comment (m-vdb) [reply rasa/core/featurizers/single_state_featurizer.py:63]: [`State`](https://github.com/RasaHQ/rasa/blob/b901d55907f9584dc4f37925e4f883c0fc7da9dc/rasa/shared/core/domain.py#L72e) is defined as a dictionary of `SubState`, which is defined as `Dict[Text, Union[Text, Tuple[Union[float, Text]]]]`, which is not actually true. The way the code uses the state: for a specific key of the `State` dict, the code expects a specific `SubState` structure. For instance:
- for `INTENT`, the `SubState` is `Dict[Text, Text]`
- for `ACTION_NAME`, the `SubState` is `TypedDict(intent=Text, text=Text, entities=List[Text])`
- for `SLOT`, the `SubState` is `Dict[Text, Union[Text, Tuple[float]]]`

`_state_features_for_attribute()` is kind of uncorrelating this dependency and assumes that because the attribute is of a certain value, the sub state will have a certain structure. Which can become risky. Does that make sense?

File: rasa/shared/core/events.py
Comment (wochinge) [reply rasa/shared/core/events.py:1215]: Yes, it's always either.  What do you think of doing something like:
```
        if self.action_name:
            return {ACTION_NAME: self.action_name}
        elif self.action_text:
            return {ACTION_TEXT: self.action_text}
       else:
           raise ValueError("{self.class.__name} does neither have {ACTION_NAME} nor {ACTION_TEXT} specified. Please check if your training data is valid."
```

This exception should never come up. However, this makes it fail loud and clear.

File: rasa/core/featurizers/single_state_featurizer.py
Comment (wochinge) [reply rasa/core/featurizers/single_state_featurizer.py:63]: Thanks, makes sense! Should we also fix the type annotation for `SubState` then? At least the `ACTION_NAME` case doesn't seem to be covered, is it?

Comment (m-vdb) [reply rasa/core/featurizers/single_state_featurizer.py:63]: honestly I've tried and I couldn't make it work without refactoring the code in `rasa/core/featurizers/single_state_featurizer.py` and I was afraid of the consequences 😬 I don't know how much effort would be required to to it properly

File: rasa/shared/core/events.py
Comment (m-vdb) [reply rasa/shared/core/events.py:1215]: yes I wanted to do that as well, but it won't fix the typing issue I'm afraid, as it's unlikely that mypy will pick it up. I'll try though

File: rasa/core/featurizers/single_state_featurizer.py
Comment (wochinge) [reply rasa/core/featurizers/single_state_featurizer.py:63]: ok, let's skip it then

**Comment by wochinge** [comment]
@m-vdb What's the state of this? Do you require another review?

**Comment by wochinge** [comment]
@m-vdb Can we wrap this up please? Happy to help btw. As this is sort of a preventive measure for potential bugs, I'd prefer to have this in asap.

**Comment by m-vdb** [comment]
will give you a status update this week, probably tomorrow 👍🏻

**Comment by m-vdb** [comment]
@wochinge I've updated the status in the PR description. Let me know how you want to proceed. If we want to merge this this week or the next, I'll need some help to implement the changes.

File: rasa/core/actions/action.py
Comment (wochinge) [reply rasa/core/actions/action.py:726]: Let's go with the exception then. How about `ActionExecutionRejection` ? The message of `ActionExecutionRejection` might need some adaption tthough.

Comment (wochinge) [reply rasa/core/actions/action.py:726]: Agree it should end up in Sentry. Tempted to just to an `assert last_user_event` as it's not supposed to ever happen. Alternatively something like a `ValueError`?

Comment (wochinge) [reply rasa/core/actions/action.py:726]: Or even better: `raise TypeError` (https://docs.python.org/3/library/exceptions.html#TypeError) ?

File: rasa/core/actions/two_stage_fallback.py
Comment (wochinge) [reply rasa/core/actions/two_stage_fallback.py:195]: How about raising a `TypeError` in case it's `null`? There should always be a `latest_message` != None

File: tests/cli/test_rasa_interactive.py
Comment (wochinge) [reply tests/cli/test_rasa_interactive.py]: Revert the change to `"train"` . Seems you already did that 💯

**Comment by wochinge** [comment]
@m-vdb Came back to the comments. If you have time to wrap it up it would be great. Otherwise I suggest to add it to the squad's board or assign it to me. Besides that I'd make it a priority for the `enable-squad` to enable the remaining checks asap (probably best to create issues for each category)

File: rasa/core/test.py
Comment (m-vdb) [reply rasa/core/test.py]: So apparently it's failing in `sklearn` with this traceback:
```
Traceback (most recent call last):
  File "/home/runner/work/rasa/rasa/.venv/bin/rasa", line 5, in <module>
    main()
  File "/home/runner/work/rasa/rasa/rasa/__main__.py", line 116, in main
    cmdline_arguments.func(cmdline_arguments)
  File "/home/runner/work/rasa/rasa/rasa/cli/test.py", line 222, in test
    run_core_test(args)
  File "/home/runner/work/rasa/rasa/rasa/cli/test.py", line 108, in run_core_test
    additional_arguments=vars(args),
  File "/home/runner/work/rasa/rasa/rasa/test.py", line 158, in test_core
    core_test(stories, _agent, out_directory=output, **kwargs)
  File "/home/runner/work/rasa/rasa/rasa/utils/common.py", line 307, in run_in_loop
    result = loop.run_until_complete(f)
  File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
  File "/home/runner/work/rasa/rasa/rasa/core/test.py", line 798, in test
    targets, predictions, output_dict=True
  File "/home/runner/work/rasa/rasa/rasa/test.py", line 331, in get_evaluation_metrics
    targets, predictions, labels=labels, output_dict=output_dict
  File "/home/runner/work/rasa/rasa/.venv/lib/python3.7/site-packages/sklearn/utils/validation.py", line 63, in inner_f
    return f(*args, **kwargs)
  File "/home/runner/work/rasa/rasa/.venv/lib/python3.7/site-packages/sklearn/metrics/_classification.py", line 1966, in classification_report
    y_type, y_true, y_pred = _check_targets(y_true, y_pred)
  File "/home/runner/work/rasa/rasa/.venv/lib/python3.7/site-packages/sklearn/metrics/_classification.py", line 83, in _check_targets
    check_consistent_length(y_true, y_pred)
  File "/home/runner/work/rasa/rasa/.venv/lib/python3.7/site-packages/sklearn/utils/validation.py", line 263, in check_consistent_length
    " samples: %r" % [int(l) for l in lengths])
ValueError: Found input variables with inconsistent numbers of samples: [51, 50]
```

Reverted my changes and updated the types of the elements in the store in this commit: 39859b4da0c

File: rasa/shared/nlu/training_data/lookup_tables_parser.py
Comment (m-vdb) [reply rasa/shared/nlu/training_data/lookup_tables_parser.py:28]: ok I think I finally get it. The typing is correct because the Markdown reader adds the lookup table file [here](https://github.com/RasaHQ/rasa/blob/8692ea37d03fa4c4244919e0ab310ed1691990a6/rasa/shared/nlu/training_data/formats/markdown.py#L106) (which is maybe questionable, but out of the scope here IMO). But if during parsing we're trying to add an item to this unloaded lookup table file, then it means that the parsing is broken. Implemented this behaviour here: 3db8c33d327

File: rasa/cli/train.py
Comment (wochinge) [inline rasa/cli/train.py:21]: this will probably be fixed by Tom [here](https://github.com/RasaHQ/rasa/pull/8141) but can we please use local imports here to avoid slow imports?

File: rasa/shared/core/trackers.py
Comment (wochinge) [inline rasa/shared/core/trackers.py:76]:
```suggestion
            LOOP_NAME: Optional[Text],
```

File: rasa/shared/core/training_data/structures.py
Comment (wochinge) [inline rasa/shared/core/training_data/structures.py:182]: 🙌🏻

File: rasa/shared/nlu/training_data/lookup_tables_parser.py
Comment (wochinge) [reply rasa/shared/nlu/training_data/lookup_tables_parser.py:28]: Why can this be a `Union[Text, List]` in the first place? What was wrong with the old type annotation?

File: rasa/core/channels/rasa_chat.py
Comment (wochinge) [inline rasa/core/channels/rasa_chat.py:77]: Is this really a `TypeError` in this case? It rather points to a wrong configuration or some problems with Rasa X to me.

File: rasa/shared/nlu/training_data/lookup_tables_parser.py
Comment (m-vdb) [reply rasa/shared/nlu/training_data/lookup_tables_parser.py:28]: I would answer this question the same way as in my previous comment, that's the way the code works now, the type annotation didn't reflect how the code works now 🤷🏻 if this answer is not good enough, can you detail a bit more what you're looking for? 😬

File: rasa/core/channels/rasa_chat.py
Comment (m-vdb) [reply rasa/core/channels/rasa_chat.py:77]: The way I understand the code: if this attribute is `None`, it's because [these lines](https://github.com/RasaHQ/rasa/blob/f89b953866d0c85a1228fca6f7852329df5cf522/rasa/core/channels/rasa_chat.py#L90-L91) failed to do their job. And actually we're all ready logging an error [here](https://github.com/RasaHQ/rasa/blob/f89b953866d0c85a1228fca6f7852329df5cf522/rasa/core/channels/rasa_chat.py#L67-L73). So maybe you're right, and this is more a problem with how the user uses rasa/rasa X, and in that case we could raise a RasaException (which wouldn't end up in Sentry)

File: rasa/shared/nlu/training_data/lookup_tables_parser.py
Comment (wochinge) [reply rasa/shared/nlu/training_data/lookup_tables_parser.py:28]: ah, I see,the markdown reader is doing that so the old type annotation was wrong 👍🏻

File: rasa/core/channels/rasa_chat.py
Comment (wochinge) [reply rasa/core/channels/rasa_chat.py:77]: True, but e.g. what if the `key` is a blank string? I'd use some `RasaException` type 👍🏻

Comment (m-vdb) [reply rasa/core/channels/rasa_chat.py:77]: 👍🏻

**Review by wochinge — APPROVED** [review]
Nice to get this in! 🚀 Great work! Do we have follow up issues for the enable squad for this?

**Comment by m-vdb** [comment]
@wochinge no we don't as I don't know how the squad wants to split the work on this yet. Maybe let's have a single placeholder issue to discuss this, and create other issues when we're ready to tackle it?

**Comment by wochinge** [comment]
Do you have some up do date statistics how often each error type is occuring? How about putting this in the squad channel? In my opinion we should have ready to go issue for refinement and then sprint planning.

**Comment by m-vdb** [comment]
let me compile this and share it

**Pull request merged**
